tools/src/main/scala/scala/scalanative/optimizer/pass/InstCombine.scala

```diff
@@ -0,0 +1,228 @@
+  }
+
+  private def simplify(inst: Inst, defop: DefOp): Seq[Inst] = {
+    val singleSimp = inst match {
+
+      case Let(z, Op.Bin(Iadd, ty, y, IVal(a))) =>
+        defop(y) match {
+          // (x + b) + a = x + (a + b)
+          case Some(Op.Bin(Iadd, _, x, IVal(b))) =>
+            Let(z, Op.Bin(Iadd, ty, x, IVal(a + b, ty)))
+
+          // (x - b) + a = x + (a - b)
+          case Some(Op.Bin(Isub, _, x, IVal(b))) =>
+            Let(z, Op.Bin(Iadd, ty, x, IVal(a - b, ty)))
+
+          case _ => inst
+        }
+
+      case Let(z, Op.Bin(Isub, ty, y, IVal(a))) =>
+        defop(y) match {
+          // (x - b) - a = x - (a + b)
+          case Some(Op.Bin(Isub, _, x, IVal(b))) =>
+            Let(z, Op.Bin(Isub, ty, x, IVal(a + b, ty)))
+
+          // (x + b) - a = x - (a - b)
+          case Some(Op.Bin(Iadd, _, x, IVal(b))) =>
+            Let(z, Op.Bin(Isub, ty, x, IVal(a - b, ty)))
+
+          case _ => inst
+        }
+
+      case Let(z, Op.Bin(Imul, ty, y, IVal(a))) =>
+        defop(y) match {
+          // (x * b) * a = x * (a * b)
+          case Some(Op.Bin(Imul, _, x, IVal(b))) =>
+            Let(z, Op.Bin(Imul, ty, x, IVal(a * b, ty)))
+
+          // (x << b) * a = x * (a * 2^b)
+          case Some(Op.Bin(Shl, _, x, IVal(b))) =>
+            Let(z, Op.Bin(Imul, ty, x, IVal(a * math.pow(2, b).toLong, ty)))
+
+          case _ => inst
+        }
+
+      case Let(n, Op.Comp(Ieq, tyIeq, compared, IVal(0))) =>
+        defop(compared) match {
+          // ((lhs xor rhs) == 0) = (lhs == rhs)
+          case Some(Op.Bin(Xor, tyXor, lhs, rhs)) if (tyIeq == tyXor) =>
+            Let(n, Op.Comp(Ieq, tyIeq, lhs, rhs))
+
+          case _ => inst
+        }
+
+      case Let(n, Op.Comp(Ine, tyIne, compared, IVal(0))) =>
+        defop(compared) match {
+          // ((lhs xor rhs) != 0) = (lhs != rhs)
+          case Some(Op.Bin(Xor, tyXor, lhs, rhs)) if (tyIne == tyXor) =>
+            Let(n, Op.Comp(Ine, tyIne, lhs, rhs))
+
+          case _ => inst
+        }
+
+      case Let(n, Op.Select(cond, thenv, elsev)) =>
+        defop(cond) match {
+          // select (c xor true) then a else b  =  select c then b else a
+          case Some(Op.Bin(Xor, _, negCond, Val.True)) =>
+            Let(n, Op.Select(negCond, elsev, thenv))
+
+          case _ => inst
+        }
+
+      case If(cond, thenp, elsep) =>
+        defop(cond) match {
+          // if (c xor true) then a else b  =  if c then b else a
+          case Some(Op.Bin(Xor, _, negCond, Val.True)) =>
+            If(negCond, elsep, thenp)
+
+          case _ => inst
+        }
+
+      case Let(n, op @ Op.Bin(_, _, lhs, rhs)) =>
+        Let(n, simplifyBin(defop(lhs), defop(rhs), op))
+
+      // Nothing
+      case _ => inst
+    }
+
+    simplifyExt(singleSimp, defop)
+  }
+
+  def simplifyBin(lhsDef: Option[Op], rhsDef: Option[Op], op: Op.Bin): Op = {
+    (lhsDef, rhsDef, op) match {
+
+      // (a * b) % a = 0
+      // (a * b) % b = 0
```

Comment:
Is this well-known? I am not convinced this will always work when `a * b` overflows.